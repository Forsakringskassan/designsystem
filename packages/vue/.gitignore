# vite dev
/src/vite-dev/app.vue
