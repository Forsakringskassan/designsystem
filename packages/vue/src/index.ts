/* include typescript declarations for vite static asset handling (e.g. `?raw`) */
/// <reference types="vite/client" />

export * from "./components";
export {
    type UseSlotUtils,
    type WithoutInstance,
    useSlotUtils,
} from "./composables";
export {
    config,
    type FKUIConfig,
    FKUIConfigButtonOrder,
    type FKUIContext,
    type MaybeWithFKUIContext,
    setRunningContext,
} from "./config";
export * from "./internal-components";
export * from "./plugins";
export {
    type ComponentUnmountEvent,
    type ComponentValidityEvent,
    type ErrorItem,
    type GroupValidityEvent,
    type ListArray,
    type ListItem,
    type UnknownItem,
    ErrorData,
    ErrorViewData,
    FormErrorList,
    MenuAction,
} from "./types";
export * from "./utils";
