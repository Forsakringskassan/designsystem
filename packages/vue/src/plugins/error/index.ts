export {
    type ErrorPluginOptions,
    default as ErrorPlugin,
    UNHANDLED_ERROR_EVENT,
} from "./ErrorPlugin";
export { default as FErrorHandlingApp } from "./FErrorHandlingApp.vue";
