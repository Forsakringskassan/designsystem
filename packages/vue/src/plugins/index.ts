export * from "./test";
export * from "./translation";
export * from "./validation";
export * from "./error";
export { FormatPlugin } from "./format";
