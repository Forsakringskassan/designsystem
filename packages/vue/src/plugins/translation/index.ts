/* istanbul ignore file */
export { type TranslateFunction } from "./translate-function";
export { TranslationMixin, TranslationPlugin } from "./TranslationPlugin";
export { useTranslate } from "./use-translate";
