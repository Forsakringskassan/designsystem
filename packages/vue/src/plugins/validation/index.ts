/* istanbul ignore file */
export { ValidationPlugin } from "./ValidationPlugin";
