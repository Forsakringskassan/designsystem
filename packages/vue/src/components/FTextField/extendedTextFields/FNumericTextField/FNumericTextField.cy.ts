import { defineComponent, type Component } from "vue";
import { FTextFieldPageObject } from "../../../../cypress";
import FNumericTextField from "./FNumericTextField.vue";

function defineTestComponent(
    textFieldTemplate: string,
    components: Record<string, Component>,
): ReturnType<typeof defineComponent> {
    return defineComponent({
        name: "TestComponent",
        template: /* HTML */ `
            <button
                data-test="toggleButton"
                type="button"
                class="button"
                @click="onClickToggleShowTextField"
            >
                Toggle showing text field
            </button>
            ${textFieldTemplate}
        `,
        components,
        data() {
            return {
                show: true,
                value: "",
            };
        },
        methods: {
            onClickToggleShowTextField(): void {
                this.show = !this.show;
            },
        },
    });
}

const textField = new FTextFieldPageObject(`[data-test="textField"]`);
const toggleButtonSelector = `[data-test="toggleButton"]`;

it("should retain correct error message when existence is toggled off and on (wrong format and too long)", () => {
    // Given
    const TestComponent = defineTestComponent(
        /* HTML */ `
            <f-numeric-text-field
                v-if="show"
                v-model="value"
                v-test="'textField'"
                v-validation.required.maxLength="{
                maxLength: { length: 3 },
            }"
            >
                Test
            </f-numeric-text-field>
        `,
        { FNumericTextField },
    );
    cy.mount(TestComponent);

    textField.enter("Wrong format and too long");
    textField.label.errorMessage().should("be.visible");
    textField.label
        .errorMessage()
        .invoke("text")
        .invoke("trim")
        .then((expectedErrorMessage) => {
            // When
            cy.get(toggleButtonSelector).click();
            textField.el().should("not.exist");
            cy.get(toggleButtonSelector).click();
            textField.el().should("be.visible");

            // Then
            textField.label.errorMessage().should("be.visible");
            textField.label
                .errorMessage()
                .invoke("text")
                .invoke("trim")
                .should("equal", expectedErrorMessage);
        });
});

it("should retain correct error message when existance is toggled off and on (wrong format)", () => {
    // Given
    const TestComponent = defineTestComponent(
        /* HTML */ `
            <f-numeric-text-field
                v-if="show"
                v-model="value"
                v-test="'textField'"
                v-validation.required.maxLength="{
                maxLength: { length: 20 },
            }"
            >
                Test
            </f-numeric-text-field>
        `,
        { FNumericTextField },
    );
    cy.mount(TestComponent);

    textField.enter("Wrong format");
    textField.label.errorMessage().should("be.visible");
    textField.label
        .errorMessage()
        .invoke("text")
        .invoke("trim")
        .then((expectedErrorMessage) => {
            // When
            cy.get(toggleButtonSelector).click();
            textField.el().should("not.exist");
            cy.get(toggleButtonSelector).click();
            textField.el().should("be.visible");

            // Then
            textField.label.errorMessage().should("be.visible");
            textField.label
                .errorMessage()
                .invoke("text")
                .invoke("trim")
                .should("equal", expectedErrorMessage);
        });
});

it("should show format error message when v-model has wrong value upon mount", () => {
    // Given
    const TestComponent = defineComponent({
        name: "TestComponent",
        template: /* HTML */ `
            <f-numeric-text-field
                v-model="value"
                v-test="'textField'"
                v-validation.required.maxLength="{
                maxLength: { length: 3 },
            }"
            >
                Test
            </f-numeric-text-field>
        `,
        components: { FNumericTextField },
        data() {
            return {
                value: "Malformatted and too long",
            };
        },
    });
    cy.mount(TestComponent);

    textField.label.errorMessage().should("be.visible");
    textField.label
        .errorMessage()
        .invoke("text")
        .invoke("trim")
        .should("equal", "Du har fyllt i ett ogiltigt tecken. Fyll i siffror.");
});
