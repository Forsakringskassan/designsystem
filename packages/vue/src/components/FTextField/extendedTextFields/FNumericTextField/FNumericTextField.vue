<script lang="ts">
import { formatNumber, parseNumber, ValidationService } from "@fkui/logic";
import { defineComponent, type PropType } from "vue";
import FTextField from "../../FTextField.vue";
import { FormatFunction, ParseFunction } from "../../index";
import { getInputElement } from "../../../../utils";
import { useTextFieldSetup } from "../../useTextFieldSetup";

function defaultFormatter(this: { decimals: number }, modelValue: number): string | undefined {
    return formatNumber(modelValue, this.decimals);
}

export default defineComponent({
    name: "FNumericTextField",
    extends: FTextField,
    props: {
        /**
         * The number of decimals to format number as.
         * @model
         */
        decimals: {
            type: Number,
            required: false,
            default: undefined,
        },
        formatter: {
            type: Function as PropType<FormatFunction<number>>,
            required: false,
            default: defaultFormatter as FormatFunction<number>,
        },
        parser: {
            type: Function as PropType<ParseFunction<number>>,
            required: false,
            default: parseNumber,
        },
    },
    setup(props) {
        return useTextFieldSetup(props);
    },
    mounted() {
        const inputElement = getInputElement(this);
        inputElement.setAttribute("inputmode", "numeric");
        inputElement.setAttribute("maxlength", "20");

        ValidationService.addValidatorsToElement(
            inputElement,
            {
                number: {},
            },
            true,
        );

        ValidationService.validateElement(inputElement);
    },
});
</script>
