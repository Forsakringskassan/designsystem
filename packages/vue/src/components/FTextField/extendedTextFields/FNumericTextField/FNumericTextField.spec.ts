import "html-validate/jest";
import { mount, VueWrapper } from "@vue/test-utils";
import { defineComponent } from "vue";
import { ValidationPlugin } from "../../../../plugins";
import FNumericTextField from "./FNumericTextField.vue";

const TestComponent = defineComponent({
    name: "TestComponent",
    components: {
        FNumericTextField,
    },
    data() {
        return {
            myModel: "" as string | number,
        };
    },
    template: "",
});

function createWrapper(
    template: string,
    myModel: string | number,
): VueWrapper<InstanceType<typeof TestComponent>> {
    return mount(TestComponent, {
        template,
        data() {
            return {
                myModel,
            };
        },
        global: {
            plugins: [ValidationPlugin],
        },
    });
}

describe("snapshots", () => {
    it("should match snapshot with label and input", () => {
        const markup = /* HTML */ `
            <f-numeric-text-field id="myField" v-model="myModel">
                My numeric field
            </f-numeric-text-field>
        `;
        const wrapper = createWrapper(markup, "");
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("v-model", () => {
    it("should update model with number value", async () => {
        const markup = /* HTML */ `
            <f-numeric-text-field
                id="myField"
                v-model="myModel"
                v-validation.decimal
            >
                My numeric field
            </f-numeric-text-field>
        `;
        const wrapper = createWrapper(markup, "");

        const input = wrapper.get("input");
        input.setValue("3,14");
        await input.trigger("blur");

        expect(wrapper.vm.$data.myModel).toBe(3.14);
    });

    it("should be able to handle zero value", async () => {
        const markup = /* HTML */ `
            <f-numeric-text-field id="myField" v-model="myModel">
                My numeric field
            </f-numeric-text-field>
        `;
        const wrapper = createWrapper(markup, "");

        const input = wrapper.get("input");
        input.setValue("0");
        await input.trigger("blur");

        expect(wrapper.vm.$data.myModel).toBe(0);
    });
});

describe("format", () => {
    it("should format with decimals when having decimal prop", async () => {
        const markup = /* HTML */ `
            <f-numeric-text-field
                id="myField"
                v-model="myModel"
                :decimals="2"
                v-validation.decimal
            >
                My numeric field
            </f-numeric-text-field>
        `;
        const wrapper = createWrapper(markup, 3);

        const input = wrapper.get("input");
        const inputElement = input.element;

        expect(inputElement.value).toBe("3,00");

        input.setValue("4,1");
        await input.trigger("blur");
        await wrapper.vm.$nextTick();

        expect(inputElement.value).toBe("4,10");
    });
});
