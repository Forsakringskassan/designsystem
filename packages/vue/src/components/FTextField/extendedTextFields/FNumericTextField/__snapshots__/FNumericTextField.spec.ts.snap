// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with label and input 1`] = `
<div
  class="text-field"
>
  <div
    class="i-width-sm-12"
  >
    <label
      class="label"
      for="myField"
    >
      
      
       My numeric field 
      
      
      
      
      <!--v-if-->
    </label>
  </div>
  <div
    class="text-field__input-wrapper i-width-sm-12"
  >
    
    
    <div
      class="text-field__icon-wrapper"
    >
      <input
        aria-invalid="false"
        class="text-field__input"
        data-validation=""
        id="myField"
        inputmode="numeric"
        maxlength="20"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    
  </div>
  <!--v-if-->
</div>
`;
